src/model/generalActions.js

```diff
@@ -190,36 +220,57 @@ const reactToGeneralAction = (model) =>
         break;
 
       case 'LEAVE_CANVAS':
-        if ((model.mode === 'FREEDRAW' || model.mode === 'POLYDRAW' || model.mode === 'RECTANGLE') 
-            && model.activityInProgress === true){
-          model.activityInProgress = false;
-          model.clicks = 0;
+        if (model.activityInProgress === false) {
```

Comment:
pls add comments here explaining this logic. This function is now too big to understand, should be broken down into logical parts